### DISCUSSION TITLE
No injected serializer when using custom normalizer that decorate the ObjectNormalizer

### DISCUSSION THREAD

**Discussion by florentpoujol** [symfony/symfony#58707] [Q&A] [answered]
Hello

I'm on Symfony 7.1.x.

The make command has generated a normalizer with a constructor like so:
```php
final class AuditLogDataNormalizer implements NormalizerInterface
{
    public function __construct(
        #[Autowire(service: 'serializer.normalizer.object')]
        private NormalizerInterface $normalizer,
    ) {
    }
// ...
}
```

The decorated normalizer is indeed `\Symfony\Component\Serializer\Normalizer\ObjectNormalizer`, but when I pass a Doctrine entity to its normalize() method it fails at the first property that has an object as value (in my example a DateTimeImmutable).

> Symfony\Component\Serializer\Exception\LogicException: Cannot normalize attribute "createdAt" because the injected serializer is not a normalizer.
> /myapp.../vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:218

The "injected serializer" is `null`.

My custom normalizer is injected into a Doctrine listener via the NormalizerInterface and no Autowire attribute.

During unit tests (KernelTestCase), I can "fix" this by manually resolving the SerializerInterface out of the container before resolving my normalizer, but obviously I shall not have to do that.  
My app do use the default services.yml config file, but it doesn't work better if I manually register the service.

So what is missing in my setup ?

Thanks !

**Comment by stof** [comment, marked as answer]
the ObjectNormalizer is implementing NormalizerAwareInterface, to be able to delegate properties to all known normalizers. But if you don't use it through the Serializer class (which is the one implement NormalizerInterface in a way knowing all normalizers and also the one handling the NormalizerAwareInterface to inject itself as the main normalizer), it won't have a main normalizer to be used for properties.

I think your Doctrine listener should not be using AuditLogDataNormalizer directly. It should use the `normalize` method of the Serializer class instead, to support using all relevant normalizers.

**Comment by florentpoujol** [comment]
Thanks, so OK it kind of makes sens: we are never supposed to use any serializer/normalizer directly by name and instead always use the SerializerInterface or NormalizerInterface.

Except in the case of normalizers depending on others like in my case, it is not a mistake to specifically ask for the ObjectNormalizer in my own normalizer.

My mistake using my normalizer directly in my service instead of depending on the general NormalizerInterface.

So actually I removed the interface from my custom normalizer, it is only used in one specific place, and I removed the Autowire attribute so that it receive `\Symfony\Component\Serializer\Serializer` instead of the `ObjectNormalizer`  and it works now.
